Update input selection process

laumair opened the pull request:
# Description

Currently, inputs are selected in a sequential way (starting from the least index with balance and stopping at the index where it satisfies the provided threshold). 

Input selection now accepts a _limit_ to restrict the number of inputs wherever necessary as is expected to work as follows:

**Case 1: When no limit is provided**
Select inputs by optimal value i.e. select inputs as few as possible. 

**Case 2: When limit is provided**
Try to select inputs by optimal value i.e. select inputs as few as possible. If selected inputs exceed limit then: 

If possible, select a set of inputs within range where _Sum_(inputs) == transaction value. This will avoid remainders. Otherwise, select a set of least size inputs within range.

Todo: Need to fix/update some tests related to [makeTransaction](https://github.com/iotaledger/trinity-wallet/blob/develop/src/shared/actions/transfers.js#L414)

Depends on https://github.com/iotaledger/trinity-wallet/pull/380

## Type of change

- Breaking change 

# How Has This Been Tested?

- Added unit test coverage

# Checklist:

_Please delete items that are not relevant._

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] For changes to `mobile` that include native code (including React Native modules): I have verified that both iOS and Android successfully build in both `Debug` and `Release` modes
- [ ] For changes to `shared`: If applicable, I have verified that my changes are implemented correctly in `desktop` and `mobile`

Commit:
Change input selection

Commit:
Merge develop into feat/input-selection

Commit:
Add coverage for #prepareInputs

Commit:
Clean up #getInputs and related utils

Commit:
Remove unused utils #getOutgoingTransfersForAddresses & #getPendingOutgoingTransfersForAddresses

Commit:
Update coverage for libs/iota/addresses

Commit:
Add coverage for #getInputs

Commit:
Fix failing tests in libs/iota/transfers

Commit:
Merge branch 'develop' into feat/input-selection

Commit:
Update test coverage for actions/transfers

rajivshah3 committed:
Merge branch 'develop' into feat/input-selection

Commit:
Merge develop into feat/input-selection

Commit:
Merge remote tracking branch feat/input-selection

Commit:
Add maxInputs property to SeedVault class and use it during input selection

rihardsgravis commented on src/desktop/src/libs/SeedStore/Keychain.js:
```diff
@@ -11,13 +11,17 @@ class Keychain {
+     * @param {number} maxInputs - Maximum number of inputs allowed per bundle
      */
-    constructor(key, accountId) {
+    constructor(key, accountId, maxInputs) {
```
I would suggest this to move as a static read-only property of the seedvault class itself (as it should only change form one seedvault type to other), see proposed codebase in the Ledger implementation.

laumair replied:
Resolved in cc51b20

Commit:
Remove unnecessary check for remainder transaction when inputs are categorised

Some bundles (e.g., MECOSAGPFIHBAJQBBX9HUJRNJEKRVNRJBHEBWYPD9H9IKBFEJZCRLLFYLLILEUHCUNYEHBEQARWNHFWBC)
are not properly categorised into inputs/outputs. This is because input transactions of such bundles have currentIndex === lastIndex.

This commit removes the unnecessary checks on inputs (restricting them to be non-remainder transactions). Categorisation of a bundle to inputs/outputs should only be based on negative/non-negative value.

 # Please enter the commit message for your changes. Lines starting

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into feat/input-selection

Commit:
Make maxInputs a read-only property of SeedVault class

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch develop into feat/input-selection

Commit:
Fix typo
